pkg/metrics/metrics.go

```diff
@@ -129,6 +129,36 @@ var (
 		Help:        "Last timestamp when we received an event",
 		ConstLabels: prometheus.Labels{"source": LabelEventSourceAPI},
 	})
+
+	// L7 statistics
+
+	// ProxyParseErrors is a count of failed parse errors on proxy
+	ProxyParseErrors = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: Namespace,
+		Name:      "policy_l7_parse_errors_total",
+		Help:      "Number of total L7 parse errors",
+	})
+
+	// ProxyForwarded is a count of all forwarded requests by proxy
+	ProxyForwarded = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: Namespace,
+		Name:      "policy_l7_forwarded_total",
+		Help:      "Number of total L7 forwarded requests/responses",
+	})
+
+	// ProxyDenied is a count of all denied requests by policy by the proxy
+	ProxyDenied = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: Namespace,
+		Name:      "policy_l7_denied_total",
+		Help:      "Number of total L7 denied requests/responses due to policy",
+	})
```

Comment:
I'm on the fence about separate `policy_l7_denied_total` and  `policy_l7_forwarded_total`. It doesn't really matter, since one can sum these two to get the total number of L7 requests. On the other hand, I would have expected the policy_l7_total to give me the total, and I could then split them out by label (`forwarded` and `denied`).
Similarly, what is the relationship with `policy_l7_recieved_total` below, and `policy_l7_parse_errors_total` above? Are some of these egress and others ingress? Those would be candidates for labels as well/instead.
This is mostly based on my reading of https://prometheus.io/docs/practices/naming/, especially the HTTP examples.

Reply:
We already collect these separate stats for Proxy and I simply exported them over. My understanding is we need even if we can derive one from the sum/total, the fact that the sums might not add up should give some clue as to if something is going wrong? This will help help understand if cilium is behaving incorrectly, so I exported all the metrics we collect.
Let me know your thoughts. Since this is more of a first step, we can always iterate and make this better.

These metrics are the sum total of both ingress/egress (see `UpdateProxyStatistics`) and are simply calculating the total of the requests/response they see.  So `policy_l7_forwarded_total` should be 2 for a legit HTTP request and corresponding response.
The GH issue did not ask for ingress/egress separation, so as a beginning step I simply exported the values we already have in the Proxy. 

`policy_l7_recieved_total` are the total number of packets received independent of whether they were denied, dropped et al.

`policy_l7_parse_errors_total` are due to errors during parsing of the requests/responses (not necessarily due to policy denial, but more like a corrupt request / something our parser does not support et al)

Reply:
@raybejjani : I actually found issues with the already present metrics. Promtool did not complain about any of the L7 proxy metrics I have added as part of this PR.

```
$ cat metrics.prom | promtool check-metrics
$ curl -s http://localhost:9090/metrics | promtool check-metrics
/prometheus # cat metrics.txt  | promtool check-metrics
cilium_endpoint_regenerations: counter metrics should have "_total" suffix
cilium_policy_import_errors: counter metrics should have "_total" suffix
cilium_endpoint_count: non-histogram and non-summary metrics should not have "_count" suffix
cilium_policy_count: non-histogram and non-summary metrics should not have "_count" suffix
/prometheus # 
```

I have raised GH #4053  to address this separately. Thanks!

Reply:
Yeah, the already present ones are bad, but they're part of the API now and I'm not sure how stable we want to keep the metrics API. In that sense, by merging this PR we are also committing to this naming scheme for these new ones.

> We already collect these separate stats for Proxy and I simply exported them over.

It's not that relevant what the proxy tracks, though. These metrics are not the internal cilium metrics, and we shouldn't assume they are (although it would be useful to export those in some debug section). These are operational metrics that an operator would use to characterise the health of the cluster or track traffic/usage patterns. In that sense, these aren't the internal state exactly, but the user-relevant equivalent. An example would be: endpoints have an internal state machine but every state may not interest the user, they might only care about `ready`, `not-ready`, `regenerating`, so we would only emit metrics with those labels (note that `not-ready` isn't even a state!).
My point was that it seems like there is one total: `policy_l7_recieved_total`. Of this total, some of the increments represent forwarded requests, some denied requests, or whatever else happens to these requests. I'm assuming these actions are exclusive. If they aren't then this applies less (we should update the descriptions to make that more clear if so). My impression is that the prevailing trend is to use labels on metrics to distinguish the states of the things being counted or measured e.g. you have one http_requests_total and each response code is a label. This allows a more top-down view into the metric: first you have a simple total, then you begin slicing it with label selectors e.g. you have http_request_duration_seconds or http_request_total, then averages, min, max, percentiles represent all http traffic. If you care about only 200 requests, you select for that only. If you want to exclude 404s because they skew the statistics, you then exclude those.
This isn't to say that multiple metrics wouldn't work in these scenarios. Older schemes did exactly that (but it's possible newer metrics systems make it hard to combine separate metrics this way, I'm not sure). I think they fell out of favour because they disallowed mixing multiple labels e.g. the http response code and also a QoS label per request. In our case, the equivalent would be how a user of these metrics wants to combine or split up this count. Are `policy_l7_forwarded_total`, `policy_l7_denied_total` and `policy_l7_received_total` always semantically different? or do they represent the same things in different ways? When I come to build a dashboard, what do I want to do with each of these independent metrics?

err, this came out as more of a rant. It was more an exercise for me to lay out my thinking.
tl;dr is that the metrics are an interface and we probably need to keep it stable: users will build alerts on them and we can't break the alerts.

Reply:
`These metrics are not the internal cilium metrics, and we shouldn't assume they are (although it would be useful to export those in some debug section)`

Actually they are. They track the parse errors /request denies / request forwarded by different proxies we use (Envoy / Kafka golang proxy).  I am not sure what you mean by `user-relevant equivalent` but the GH issue #4010 specially asked for these specific metrics:
```
Rejected L7 (HTTP/Kafka) due to parse errors.
Forwarded L7 (HTTP/Kafka)
Denied L7 (HTTP/Kafka) due to policy
```
And these are the exact metrics that are being exposed in this PR.
I think there is some confusion about why these are not internal cilium metrics.
Also, there is a subtle difference between `policy_l7_forwarded_total` , `policy_l7_denied_total` and `policy_l7_received_total`. We can obviously derive one from other, but I think the main purpose(use-case which Amey is working on) was to find out if cilium is misbehaving in  any of the data paths. Again maybe this does not fall under standard prometheus usage and maybe we can discuss your concerns offline and I can be more mindful about your concerns next time. Also, I do agree that we can aggregate these with the help of relevant labels, but my initial rationale was that these 3 are different metrics and should be exposed nevertheless.

Reply:
You're correct. I should have commented on the PR when it was first made instead of the PR. Sorry.